CRM-20450: Fix invoice math for partial payments.

KarinG opened the pull request:
See screenshot:

![invoicemathpartialpaymentsiswrong](https://cloud.githubusercontent.com/assets/5340555/25208988/527df0ba-2535-11e7-9274-0afa64a04765.png)

 OK the root issue appears to be in this line of code:
 `$amountDue = ($input['amount'] - $input['amount']);`
[looks like it was a placeholder and never got worked out];

This PR also fixes https://issues.civicrm.org/jira/browse/CRM-19837

Re: /Upgrade -> do I simply copy the entire new version of the tpl into 4.7.19.msg_template/message_templates?

---

 * [CRM-20450: Fix invoice math for partial payments](https://issues.civicrm.org/jira/browse/CRM-20450)

seamuslee001 commented:
@KarinG I believe you do and just make sure something calls the templates in 4.7.19.mysql.tpl

seamuslee001 commented:
@KarinG  Karin can you please rebase this against the current master

KarinG commented:
@seamuslee01 - it says no merge conflicts?

seamuslee001 commented:
@KarinG in Jenkins it is showing

```
[https://github.com/civicrm/civicrm-packages.git]
In "sites/all/modules/civicrm/", rename "master" to "backup-master-20170420012230-9".
In "sites/all/modules/civicrm/", create "master" using "origin/master".
In "sites/all/modules/civicrm/", apply "https://github.com/civicrm/civicrm-core/pull/10190" on top of "be4c4a853aab55c4022c8cb66b9cbd45a609e00e".

                                                                                      
  [GitScan\Exception\ProcessErrorException]                                           
  Process failed:                                                                     
  [[ COMMAND: git apply --check --ignore-whitespace ]]                                
  [[ CWD: /home/jenkins/buildkit/build/core-10190-9214s/sites/all/modules/civicrm ]]  
  [[ EXIT CODE: 1 ]]                                                                  
  [[ STDOUT ]]                                                                        
  [[ STDERR ]]                                                                        
  error: patch failed: CRM/Contribute/Form/AdditionalPayment.php:101                  
  error: CRM/Contribute/Form/AdditionalPayment.php: patch does not apply
````

eileenmcnaughton commented:
@KarinG  I think that is correct re the upgrade method

My notes here are that the code looks good & is tightly constrained - ie. should only affect the invoice screen, which you have provided screen shots of - so I feel pretty comfortable.

@agileware @JoeMurray @guanhuan do any of you have any input as I believe you all have some skin in this functionality

KarinG commented:
Thanks @eileenmcnaughton - For sake of completeness - here's an after fix screen - showing the correct Math on the invoice:

![screen shot 2017-04-20 at 9 55 31 pm](https://cloud.githubusercontent.com/assets/5340555/25262267/308b1412-2614-11e7-9197-d23dc880f05e.png)

Will try find time to get that message upgrade in tomorrow!

KarinG commented:
Ha! I just fixed all the 'enitity' typos - while I noticed them. A bit of spring cleaning.

KarinG commented:
Ok I've added the entire new message template in: CRM/Upgrade/4.7.19.msg_template/message_templates/contribution_invoice_receipt_html.tpl

When I look at 4.7.19.mysql.tpl in CRM/Upgrade/Incremental/sql -> it already calls:
{include file='../CRM/Upgrade/4.7.19.msg_template/civicrm_msg_template.tpl'}

Which in turn checks the directory that I just added the contribution_invoice_receipt_html.tpl to. 

That should do it! I hope.

eileenmcnaughton commented:
@KarinG we've cut 4.7.19 now so if it's against master it needs to be 4.7.20 :-(

KarinG commented:
Coleman told me to get this done by Monday - and I did 😀

eileenmcnaughton commented:
Hmm the rc was due to be cut on the 19th - although I think it was a day late being cut

seamuslee001 commented:
Just opened the following PR https://github.com/civicrm/civicrm-core/pull/10222 to go against the RC branch

seamuslee001 commented:
@KarinG is there a text version of the message template that also needs upgrading?

KarinG commented:
@seamuslee0001 - No there is no text version;

seamuslee001 commented:
@KarinG cool just thought to check

KarinG commented:
Yeah - no worries - thanks for opening a branch against the RC - if this gets in I may even upgrade a project to 4.7 😀

KarinG commented:
PS @seamuslee001 - would you be able to test the upgrade part? It's the first upgrade of this kind that I've had to do..

seamuslee001 commented:
sure

KarinG commented:
Thank you!

eileenmcnaughton commented:
Ok - I added unit testing to this in a separate PR, which was a bigger issue for me than the late-to-rc in this case. Issue closed. all done. Yay Thanks @KarinG @adixon @seamuslee001
